### PR TITLE
bump go to 1.22

### LINKED ISSUE #187: bump go to 1.22
Needed by #183 ~#186~

https://github.com/ionos-cloud/cluster-api-provider-proxmox/labels/go1.22

### PR DESCRIPTION
**Pull request by wikkyk**
*Issue #, if available:*
fix #187

+mass dep update:
k8s/api* to 0.30
controller-gen to 0.15.0 (0.14.0 has the go 1.22 panic fix)
controller-runtime to 0.18.0 (dependency)

Regenerated crds using new controller-gen. Note config/rbac/role.yaml and config/webhook/manifest.yaml.

### CODE REVIEW

**Comment by wikkyk** [comment]
obsoletes #190

**Comment by wikkyk** [comment]
#177

**Comment by wikkyk** [comment]
I will squash this PR into a single commit because none of commits in the PR work without all the other commits in the PR.
The only exception is the grype commit for easy revert later.

File: go.mod [hunk 1]
```diff
@@ -1,8 +1,8 @@
 module github.com/ionos-cloud/cluster-api-provider-proxmox
 
-go 1.21
+go 1.22.2
 
-replace sigs.k8s.io/cluster-api => sigs.k8s.io/cluster-api v1.6.3
+replace sigs.k8s.io/cluster-api => sigs.k8s.io/cluster-api v1.7.0-rc.0.0.20240424194424-783624c51bee
```
[the remark this answers is not in the document]
Comment (mcbenjemaa) [reply go.mod:5]: please use `v1.7.1`

[the remark this answers is not in the document]
Comment (wikkyk) [reply go.mod:5]: This commit is actually newer than the `v1.7.1` tag, `go mod` just miscalculated the version number.
We need this specific commit as it introduces support for controller-runtime v0.18.0. Unless we come up with a better fix, we'll have to wait with releasing v0.5.0 until capi v1.7.2 is released which will contain this commit.

**Review by mcbenjemaa — COMMENTED** [review]
Please check why,
targets make generate and manifests are not working.

**Comment by wikkyk** [comment]
@mcbenjemaa you need to be more specific - what doesn't work?

**Comment by mcbenjemaa** [comment]
None of the providers moved to 1.22
cause they follow capi:

- https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/main/go.mod
- https://github.com/kubernetes-sigs/cluster-api-provider-gcp/blob/main/go.mod
- https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/blob/main/go.mod
- https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/go.mod
- https://github.com/kubernetes-sigs/cluster-api-provider-packet/blob/main/go.mod
- https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/blob/main/go.mod

I recommend that we wait for the v1.8 release. To be compliant with CAPI.

**Comment by wikkyk** [comment]
capi 1.7.2 still does not support controller-runtime 0.18

**Comment by wikkyk** [comment]
Rebased for v0.5.0.

**Comment by mcbenjemaa** [comment]
We can rework this now.

**Comment by wikkyk** [comment]
Ready for review.

File: go.mod [hunk 2]
```diff
@@ -1,8 +1,10 @@
 module github.com/ionos-cloud/cluster-api-provider-proxmox
 
-go 1.21
+go 1.22.0
```
Comment (mcbenjemaa) [inline go.mod:3]: It's better to set `go 1.22.2` and remove toolchain

Comment (wikkyk) [reply go.mod:3]: indeed, pushing the change now.

**Comment by wikkyk** [comment]
e2e passed, pushing cosmetic changes to go.mod

**Review by mcbenjemaa — APPROVED** [review]
LGTM

**Pull request merged**